File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(candidates);
        List<Integer> curr = new ArrayList<>();
        backtrack(result, curr, candidates, target, 0);
        return result;
    }
    
    public void backtrack(List<List<Integer>> result, List<Integer> curr, int[] candidates, int target, int index){
        if(target == 0 && curr.size()>0){
            result.add(new ArrayList<>(curr));
            return;
        }
        if(target < 0){
            return;
        }
        for(int i=index;i<candidates.length;i++){
            if(i > index && candidates[i] == candidates[i-1]){
                continue;
            }
            curr.add(candidates[i]);
            backtrack(result, curr, candidates, target-candidates[i], i+1);
            curr.remove(curr.size()-1);
        }
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        def backtrack(remaining, index, curr):
            if remaining == 0 and len(curr) > 0:
                result.append(curr.copy())
                return
            if remaining < 0:
                return
            for i in range(index, len(candidates)):
                if i > index and candidates[i] == candidates[i-1]:
                    continue
                curr.append(candidates[i])
                backtrack(remaining - candidates[i], i+1, curr)
                curr.pop()
        candidates.sort()
        backtrack(target, 0, [])
        return result
            